homeassistant/components/bluesound/media_player.py

```diff
@@ -829,83 +812,118 @@ def is_master(self):
 
-        bluesound_group = []
+        slaves = []
+        slave_node = self._sync_status.get("slave")
+        nodes = slave_node if isinstance(slave_node, list) else [slave_node]
 
-        device_group = self._group_name.split("+")
+        for slave in nodes:
+            slave_id = slave.get("@id")
+            slave_port = slave.get("@port")
+            slaves.append(f"{slave_id}:{slave_port}")
 
-        sorted_entities = sorted(
-            self._hass.data[DATA_BLUESOUND],
-            key=lambda entity: entity.is_master,
-            reverse=True,
+        bluesound_group: list[BluesoundPlayer] = []
+        bluesound_group.append(self)
+
+        bluesound_group.extend(
+            device for device in self.hass.data[DATA_BLUESOUND] if device.id in slaves
         )
-        bluesound_group = [
-            entity.name
-            for entity in sorted_entities
-            if entity.bluesound_device_name in device_group
-        ]
 
         return bluesound_group
 
-    async def async_unjoin(self):
-        """Unjoin the player from a group."""
-        if self._master is None:
-            return
+    async def async_join_players(self, group_members: list[str]) -> None:
+        """Join `group_members` as a player group with the current player."""
+
+        devices = [
+            device
+            for device in self.hass.data[DATA_BLUESOUND]
+            if device.entity_id in group_members
+        ]
+
+        master: BluesoundPlayer
+        if self.is_master:
+            master = self
+        elif self.is_grouped:
+            master = self.master
+        else:
+            # try to be smart and check if any other devices is already in group
+            other_masters = [device for device in devices if device.is_master]
+            if len(other_masters) > 0:
+                master = other_masters[0]
+                devices.remove(master)
+                devices.append(self)
+            else:
+                other_slaves = [device for device in devices if device.is_grouped]
+                if len(other_slaves) > 0:
+                    master = other_slaves[0].master
+                    devices.append(self)
+                    devices.remove(other_slaves[0])
+                    _LOGGER.warning("5 %s", devices)
+                else:
+                    master = self
 
-        _LOGGER.debug("Trying to unjoin player: %s", self.id)
-        await self._master.async_remove_slave(self)
+        await master.add_slaves(devices)
 
-    async def async_add_slave(self, slave_device):
+    async def add_slaves(self, slave_devices):
         """Add slave to master."""
+
+        slaves = ",".join([entity.ip for entity in slave_devices])
+        ports = ",".join([str(entity.port) for entity in slave_devices])
+
         return await self.send_bluesound_command(
-            f"/AddSlave?slave={slave_device.host}&port={slave_device.port}"
+            f"/AddSlave?slaves={slaves}&ports={ports}"
```

Comment:
This integration needs a 3rd party library before we change any code that involves protocol details like this.

Reply:
Yes, I know I will try to fix this later. I have alot of other changes that I want to do but I'm waiting with those until I have the time to make the 3rd party library.
This is only to fix compatibility with current function, I'm not actually adding anything new.

Reply:
We're touching protocol code. We don't allow that. The integration needs the 3rd party library first.